File: src/org/lineageos/setupwizard/BaseSetupWizardActivity.java
/*
 * SPDX-FileCopyrightText: 2016 The CyanogenMod Project
 * SPDX-FileCopyrightText: 2017-2024 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.setupwizard;

import static android.view.View.INVISIBLE;

import static androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult;

import static com.google.android.setupcompat.util.ResultCodes.RESULT_SKIP;

import static org.lineageos.setupwizard.SetupWizardApp.LOGV;

import android.annotation.NonNull;
import android.annotation.Nullable;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import androidx.activity.OnBackPressedCallback;
import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContract;
import androidx.appcompat.app.AppCompatActivity;

import com.android.settingslib.Utils;

import com.google.android.setupcompat.util.WizardManagerHelper;
import com.google.android.setupdesign.GlifLayout;
import com.google.android.setupdesign.transition.TransitionHelper;
import com.google.android.setupdesign.util.ThemeHelper;

import org.lineageos.setupwizard.NavigationLayout.NavigationBarListener;
import org.lineageos.setupwizard.util.SetupWizardUtils;

public abstract class BaseSetupWizardActivity extends AppCompatActivity implements
        NavigationBarListener {

    public static final String TAG = BaseSetupWizardActivity.class.getSimpleName();
    public static final int DEFAULT_TRANSITION = TransitionHelper.TRANSITION_FADE_THROUGH;

    private NavigationLayout mNavigationBar;

    private final ActivityResultLauncher<Intent> mNextIntentResultLauncher =
            registerForActivityResult(
                    new StartDecoratedActivityForResult(),
                    BaseSetupWizardActivity.this::onNextIntentResult);

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        if (LOGV) {
            logActivityState("onCreate savedInstanceState=" + savedInstanceState);
        }
        super.onCreate(savedInstanceState);
        initLayout();
        mNavigationBar = getNavigationBar();
        if (mNavigationBar != null) {
            mNavigationBar.setNavigationBarListener(this);
        }
        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                if (LOGV) {
                    Log.v(TAG, "handleOnBackPressed()");
                }
                finishAction(RESULT_CANCELED, new Intent().putExtra("onBackPressed", true));
            }
        });
        // Apply default transition, to take effect whenever leaving this activity.
        applyForwardTransition();
    }

    @Override
    protected void onStart() {
        if (LOGV) {
            logActivityState("onStart");
        }
        super.onStart();
    }

    @Override
    protected void onRestart() {
        if (LOGV) {
            logActivityState("onRestart");
        }
        super.onRestart();
    }

    @Override
    protected void onResume() {
        if (LOGV) {
            logActivityState("onResume");
        }
        super.onResume();
    }

    @Override
    protected void onPause() {
        if (LOGV) {
            logActivityState("onPause");
        }
        super.onPause();
    }

    @Override
    protected void onStop() {
        if (LOGV) {
            logActivityState("onStop");
        }
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        if (LOGV) {
            logActivityState("onDestroy");
        }
        super.onDestroy();
    }

    @Override
    public void onAttachedToWindow() {
        if (LOGV) {
            logActivityState("onAttachedToWindow");
        }
        super.onAttachedToWindow();
    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        if (LOGV) {
            Log.v(TAG, "onRestoreInstanceState(" + savedInstanceState + ")");
        }
        super.onRestoreInstanceState(savedInstanceState);
        int currentId = savedInstanceState.getInt("currentFocus", -1);
        if (currentId != -1) {
            View view = findViewById(currentId);
            if (view != null) {
                view.requestFocus();
            }
        }
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        View current = getCurrentFocus();
        outState.putInt("currentFocus", current != null ? current.getId() : -1);
        if (LOGV) {
            Log.v(TAG, "onSaveInstanceState(" + outState + ")");
        }
    }

    /**
     * @return The navigation bar instance in the layout, or null if the layout does not have a
     * navigation bar.
     */
    public NavigationLayout getNavigationBar() {
        final View view = findViewById(R.id.navigation_bar);
        return view instanceof NavigationLayout ? (NavigationLayout) view : null;
    }

    public final void setNextAllowed(boolean allowed) {
        if (mNavigationBar != null) {
            mNavigationBar.getNextButton().setEnabled(allowed);
        }
    }

    protected void onNextPressed() {
        nextAction(RESULT_OK);
    }

    protected void onSkipPressed() {
        nextAction(RESULT_SKIP);
    }

    protected final void setNextText(int resId) {
        if (mNavigationBar != null) {
            mNavigationBar.getNextButton().setText(resId);
        }
    }

    public Button getNextButton() {
        return mNavigationBar.getNextButton();
    }

    protected final void setSkipText(int resId) {
        if (mNavigationBar != null) {
            mNavigationBar.getSkipButton().setText(resId);
        }
    }

    protected final void hideNextButton() {
        if (mNavigationBar != null) {
            final Button next = mNavigationBar.getNextButton();
            next.setVisibility(INVISIBLE);
        }
    }

    public void onNavigateNext() {
        onNextPressed();
    }

    public void onSkip() {
        onSkipPressed();
    }

    protected final void onSetupStart() {
        if (SetupWizardUtils.isOwner()) {
            tryEnablingWifi();
        }
    }

    public void finish() {
        if (LOGV) {
            Log.v(TAG, "finish");
        }
        super.finish();
    }

    protected final void finishAction(int resultCode) {
        finishAction(resultCode, null);
    }

    protected final void finishAction(int resultCode, Intent data) {
        if (resultCode != RESULT_CANCELED) {
            nextAction(resultCode, data);
            finish();
        } else {
            setResult(resultCode, data);
            finish();
            applyBackwardTransition();
        }
    }

    public final void nextAction(int resultCode) {
        nextAction(resultCode, null);
    }

    protected final void nextAction(int resultCode, Intent data) {
        if (LOGV) {
            Log.v(TAG, "nextAction resultCode=" + resultCode +
                    " data=" + data + " this=" + this);
        }
        if (resultCode == RESULT_CANCELED) {
            throw new IllegalArgumentException("Cannot call nextAction with RESULT_CANCELED");
        }
        setResult(resultCode, data);
        Intent intent = WizardManagerHelper.getNextIntent(getIntent(), resultCode, data);
        mNextIntentResultLauncher.launch(intent);
    }

    /** Adorn the Intent with Setup Wizard-related extras. */
    protected Intent decorateIntent(Intent intent) {
        return intent
                .putExtra(WizardManagerHelper.EXTRA_IS_FIRST_RUN, isFirstRun())
                .putExtra(WizardManagerHelper.EXTRA_IS_SETUP_FLOW, true)
                .putExtra(WizardManagerHelper.EXTRA_THEME, ThemeHelper.THEME_GLIF_V4);
    }

    @Override
    public void startActivity(Intent intent) {
        super.startActivity(decorateIntent(intent));
    }

    protected void onNextIntentResult(@NonNull ActivityResult activityResult) {
        int resultCode = activityResult.getResultCode();
        Intent data = activityResult.getData();
        if (LOGV) {
            StringBuilder append = new StringBuilder().append("onNextIntentResult(")
                    .append(resultCode).append(", ");
            Bundle extras = null;
            if (data != null) {
                extras = data.getExtras();
            }
            Log.v(TAG, append.append(extras).append(")").toString());
        }
    }

    protected final boolean tryEnablingWifi() {
        WifiManager wifiManager = getSystemService(WifiManager.class);
        return wifiManager != null && wifiManager.setWifiEnabled(true);
    }

    private boolean isFirstRun() {
        return true;
    }

    protected final void logActivityState(String prefix) {
        Log.v(TAG, prefix + " isResumed=" + isResumed() + " isFinishing=" +
                isFinishing() + " isDestroyed=" + isDestroyed());
    }

    private void initLayout() {
        if (getLayoutResId() != -1) {
            setContentView(getLayoutResId());
        }
        if (getTitleResId() != -1) {
            final CharSequence headerText = TextUtils.expandTemplate(getText(getTitleResId()));
            getGlifLayout().setHeaderText(headerText);
        }
        if (getIconResId() != -1) {
            final GlifLayout layout = getGlifLayout();
            final Drawable icon = getDrawable(getIconResId()).mutate();
            icon.setTintList(Utils.getColorAccent(layout.getContext()));
            layout.setIcon(icon);
        }
    }

    protected GlifLayout getGlifLayout() {
        return requireViewById(R.id.setup_wizard_layout);
    }

    protected int getLayoutResId() {
        return -1;
    }

    protected int getTitleResId() {
        return -1;
    }

    protected int getIconResId() {
        return -1;
    }

    protected void applyForwardTransition() {
        TransitionHelper.applyForwardTransition(this, DEFAULT_TRANSITION, true);
    }

    protected void applyBackwardTransition() {
        TransitionHelper.applyBackwardTransition(BaseSetupWizardActivity.this,
                DEFAULT_TRANSITION, true);
    }

    protected final class StartDecoratedActivityForResult
            extends ActivityResultContract<Intent, ActivityResult> {

        private final StartActivityForResult mWrappedContract = new StartActivityForResult();

        @NonNull
        @Override
        public Intent createIntent(@NonNull Context context, @NonNull Intent intent) {
            return decorateIntent(mWrappedContract.createIntent(context, intent));
        }

        @Override
        public ActivityResult parseResult(int resultCode, @Nullable Intent result) {
            return mWrappedContract.parseResult(resultCode, result);
        }
    }
}

File: app/src/main/java/org/lineageos/setupwizard/base/BaseSetupWizardActivity.kt
/*
 * SPDX-FileCopyrightText: 2016 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.setupwizard.base

import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.net.wifi.WifiManager
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.view.View.INVISIBLE
import android.widget.Button
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import com.android.settingslib.Utils
import com.google.android.setupcompat.util.ResultCodes.RESULT_SKIP
import com.google.android.setupcompat.util.WizardManagerHelper
import com.google.android.setupdesign.GlifLayout
import com.google.android.setupdesign.transition.TransitionHelper
import com.google.android.setupdesign.util.ThemeHelper
import org.lineageos.setupwizard.R
import org.lineageos.setupwizard.SetupWizardApp.Companion.LOGV
import org.lineageos.setupwizard.util.SetupWizardUtils
import org.lineageos.setupwizard.widget.NavigationLayout
import org.lineageos.setupwizard.widget.NavigationLayout.NavigationBarListener

abstract class BaseSetupWizardActivity : AppCompatActivity(), NavigationBarListener {

    private var navigationBar: NavigationLayout? = null
    private lateinit var nextIntentResultLauncher: ActivityResultLauncher<Intent>

    override fun onCreate(savedInstanceState: Bundle?) {
        if (LOGV) {
            logActivityState("onCreate savedInstanceState=$savedInstanceState")
        }
        super.onCreate(savedInstanceState)
        nextIntentResultLauncher =
            registerForActivityResult(StartDecoratedActivityForResult(), this::onNextIntentResult)
        initLayout()
        navigationBar = findViewById<View>(R.id.navigation_bar) as? NavigationLayout
        navigationBar?.setNavigationBarListener(this)
        onBackPressedDispatcher.addCallback(
            this,
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    if (LOGV) {
                        Log.v(TAG, "handleOnBackPressed()")
                    }
                    finishAction(RESULT_CANCELED, Intent().putExtra("onBackPressed", true))
                }
            },
        )
        // Apply default transition, to take effect whenever leaving this activity.
        applyForwardTransition()
    }

    override fun onStart() {
        if (LOGV) logActivityState("onStart")
        super.onStart()
    }

    override fun onRestart() {
        if (LOGV) logActivityState("onRestart")
        super.onRestart()
    }

    override fun onResume() {
        if (LOGV) logActivityState("onResume")
        super.onResume()
    }

    override fun onPause() {
        if (LOGV) logActivityState("onPause")
        super.onPause()
    }

    override fun onStop() {
        if (LOGV) logActivityState("onStop")
        super.onStop()
    }

    override fun onDestroy() {
        if (LOGV) logActivityState("onDestroy")
        super.onDestroy()
        nextIntentResultLauncher.unregister()
    }

    override fun onAttachedToWindow() {
        if (LOGV) logActivityState("onAttachedToWindow")
        super.onAttachedToWindow()
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        if (LOGV) {
            Log.v(TAG, "onRestoreInstanceState($savedInstanceState)")
        }
        super.onRestoreInstanceState(savedInstanceState)
        val currentId = savedInstanceState.getInt("currentFocus", -1)
        if (currentId != -1) {
            findViewById<View>(currentId)?.requestFocus()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("currentFocus", currentFocus?.id ?: -1)
        if (LOGV) {
            Log.v(TAG, "onSaveInstanceState($outState)")
        }
    }

    fun setNextAllowed(allowed: Boolean) {
        navigationBar?.nextButton?.isEnabled = allowed
    }

    protected open fun onNextPressed() {
        nextAction(RESULT_OK)
    }

    protected open fun onSkipPressed() {
        nextAction(RESULT_SKIP)
    }

    protected fun setNextText(resId: Int) {
        navigationBar?.nextButton?.setText(resId)
    }

    val nextButton: Button
        get() = navigationBar!!.nextButton

    protected fun setSkipText(resId: Int) {
        navigationBar?.skipButton?.setText(resId)
    }

    protected fun hideNextButton() {
        navigationBar?.nextButton?.visibility = INVISIBLE
    }

    override fun onNavigateNext() {
        onNextPressed()
    }

    override fun onSkip() {
        onSkipPressed()
    }

    protected fun onSetupStart() {
        if (SetupWizardUtils.isOwner()) {
            tryEnablingWifi()
        }
    }

    override fun finish() {
        if (LOGV) {
            Log.v(TAG, "finish")
        }
        super.finish()
    }

    protected fun finishAction(resultCode: Int, data: Intent? = null) {
        if (resultCode != RESULT_CANCELED) {
            nextAction(resultCode, data)
            finish()
        } else {
            setResult(resultCode, data)
            finish()
            applyBackwardTransition()
        }
    }

    fun nextAction(resultCode: Int, data: Intent? = null) {
        if (LOGV) {
            Log.v(TAG, "nextAction resultCode=$resultCode data=$data this=$this")
        }
        require(resultCode != RESULT_CANCELED) { "Cannot call nextAction with RESULT_CANCELED" }
        setResult(resultCode, data)
        val nextIntent = WizardManagerHelper.getNextIntent(intent, resultCode, data)
        nextIntentResultLauncher.launch(nextIntent)
    }

    /** Adorn the Intent with Setup Wizard-related extras. */
    protected open fun decorateIntent(intent: Intent): Intent =
        intent
            .putExtra(WizardManagerHelper.EXTRA_IS_FIRST_RUN, isFirstRun())
            .putExtra(WizardManagerHelper.EXTRA_IS_SETUP_FLOW, true)
            .putExtra(WizardManagerHelper.EXTRA_THEME, ThemeHelper.THEME_GLIF_EXPRESSIVE)

    override fun startActivity(intent: Intent) {
        super.startActivity(decorateIntent(intent))
    }

    protected open fun onNextIntentResult(activityResult: ActivityResult) {
        val resultCode = activityResult.resultCode
        val data = activityResult.data
        if (LOGV) {
            Log.v(TAG, "onNextIntentResult($resultCode, ${data?.extras})")
        }
    }

    protected fun tryEnablingWifi(): Boolean =
        getSystemService(WifiManager::class.java)?.setWifiEnabled(true) ?: false

    private fun isFirstRun(): Boolean = true

    protected fun logActivityState(prefix: String) {
        Log.v(TAG, "$prefix isResumed=$isResumed isFinishing=$isFinishing isDestroyed=$isDestroyed")
    }

    private fun initLayout() {
        if (layoutResId != -1) {
            setContentView(layoutResId)
        }
        if (titleResId != -1) {
            val headerText = TextUtils.expandTemplate(getText(titleResId))
            glifLayout.setHeaderText(headerText)
        }
        if (iconResId != -1) {
            val layout = glifLayout
            val icon: Drawable = getDrawable(iconResId)!!.mutate()
            icon.setTintList(Utils.getColorAccent(layout.context))
            layout.setIcon(icon)
        }
    }

    protected val glifLayout: GlifLayout
        get() = requireViewById(R.id.setup_wizard_layout)

    protected open val layoutResId: Int = -1

    protected open val titleResId: Int = -1

    protected open val iconResId: Int = -1

    protected open fun applyForwardTransition() {
        TransitionHelper.applyForwardTransition(this, DEFAULT_TRANSITION, true)
    }

    protected open fun applyBackwardTransition() {
        TransitionHelper.applyBackwardTransition(this, DEFAULT_TRANSITION, true)
    }

    protected inner class StartDecoratedActivityForResult :
        ActivityResultContract<Intent, ActivityResult>() {

        private val wrappedContract = StartActivityForResult()

        override fun createIntent(context: Context, input: Intent): Intent =
            decorateIntent(wrappedContract.createIntent(context, input))

        override fun parseResult(resultCode: Int, intent: Intent?): ActivityResult =
            wrappedContract.parseResult(resultCode, intent)
    }

    companion object {
        private const val TAG = "BaseSetupWizardActivity"
        const val DEFAULT_TRANSITION = TransitionHelper.TRANSITION_FADE_THROUGH
    }
}
